File: solutions/c/rail-fence-cipher/3/rail_fence_cipher.c
#include "rail_fence_cipher.h"

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

static char* subst(const char* in, size_t rails, bool decode) {
    size_t const len = strlen(in);
    size_t const max_step = 2 * (rails - 1);

    char* out = calloc(1, len + 1);
    size_t cipher_index = 0;
    for (size_t r = 0; r < rails; ++r) {
        size_t step = 2 * r;
        for (size_t plain_index = r; plain_index < len; plain_index += step) {
            if (decode) {
                out[plain_index] = in[cipher_index];
            } else {
                out[cipher_index] = in[plain_index];
            }
            ++cipher_index;
            if (step != max_step) step = max_step - step;
        }
    }

    return out;
}

char* encode(const char* plain, size_t rails) {
    return subst(plain, rails, false);
}

char* decode(const char* cipher, size_t rails) {
    return subst(cipher, rails, true);
}

File: solutions/haskell/rail-fence-cipher/3/src/RailFenceCipher.hs
{-# LANGUAGE BlockArguments #-}

module RailFenceCipher (encode, decode) where

import Control.Monad.ST (runST)
import Data.Foldable (for_, toList)
import Data.Functor ((<&>))
import Data.List (unfoldr)
import Data.Vector (freeze, (!))
import Data.Vector qualified as Vector
import Data.Vector.Mutable (new, write)

data Codec = Encode | Decode

encode :: Int -> String -> String
encode = subst Encode

decode :: Int -> String -> String
decode = subst Decode

subst :: Codec -> Int -> String -> String
subst codec rails inputStr =
    runST do
        output <- new $ length input
        for_ (zip cipherIxs plainIxs) \(cipherIx, plainIx) ->
            case codec of
                Decode -> write output plainIx $ input ! cipherIx
                Encode -> write output cipherIx $ input ! plainIx
        freeze output <&> toList
  where
    input = Vector.fromList inputStr
    plainIxs = makePlainIndices rails $ length input
    cipherIxs = [0 ..]

makePlainIndices :: Int -> Int -> [Int]
makePlainIndices rails messageLength = concatMap railPlainIxs [0 .. rails - 1]
  where
    maxStep = 2 * (rails - 1)

    nextStep s = if s /= maxStep then maxStep - s else s

    railPlainIxs rail
        | rails == 1 = takeWhile (< messageLength) [rail ..]
        | otherwise = unfoldr step (rail, nextStep (2 * rail))

    step (ix, s) =
        (ix < messageLength) `thenJust` (ix, (ix + s, nextStep s))

thenJust :: Bool -> a -> Maybe a
thenJust False _ = Nothing
thenJust True x = Just x
